### PR TITLE
Refactor/course controller rework

### PR DESCRIPTION
**Pull request by tomwim**
### Description
Rework of the course controller logic to separate logic from UI. The course controller now is intended to only spawn a course controller which is independent of the course controller menu. This allows the user to set a menu for the application independent of the course controller and rig. 
Additionally, the initialization of the Course now is done by the spawned Course Controller and not in the Menu anymore.

### Type of change
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected) ❓ 
- [X] This change requires a documentation update

### How Has This Been Tested?
Extended and created new CourseControllers to use a menu.
Tested for Default and Standalone Menu.
Change to custom prefabs.

### Aditional Notes
Might be a breaking change because prefabs changed. Not sure how to handle this.

### CODE REVIEW

File: Runtime/CourseController/Controllers/BaseCourseController.cs [hunk 1]
```diff
@@ -31,5 +35,15 @@ public virtual List<Type> GetRequiredSetupComponents()
         {
             return new List<Type>();
         }
+
+        /// <summary>
+        /// Handles post-setup logic.
+        /// Should be called after all components are added and object is initialized.
+        /// </summary>
+        /// <param name="courseControllerObject"></param>
+        public virtual void SetupDone(GameObject courseControllerObject)
```
Comment (SimonTheSourcerer) [inline Runtime/CourseController/Controllers/BaseCourseController.cs:44]:
```suggestion
        /// </summary>
        public virtual void SetupDone(GameObject courseControllerObject)
```

File: Runtime/CourseController/Controllers/ICourseController.cs [hunk 2]
```diff
@@ -30,5 +30,7 @@ public interface ICourseController
         /// </summary>
         /// <returns>List of component types.</returns>
         List<Type> GetRequiredSetupComponents();
+
+        void SetupDone(GameObject courseControllerObject);
```
Comment (SimonTheSourcerer) [inline Runtime/CourseController/Controllers/ICourseController.cs:34]: xml doc

File: Runtime/CourseController/InitCourseOnSceneLoad.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+namespace Innoactive.Creator.UX
+{
+    /// <summary>
+    /// Initializes the <see cref="CourseRunner"/> with the current selected course on scene start.
+    /// </summary>
+    public class InitCourseOnSceneLoad : MonoBehaviour
+    {
+        private void OnEnable()
+        {
+            InitTraining();
```
Comment (SimonTheSourcerer) [inline Runtime/CourseController/InitCourseOnSceneLoad.cs:15]: You can select the language with the menu loaded, is this way here still supporting switching the language?

File: Runtime/CourseController/CourseMenuSpawner.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            }
+            else
+            {
+                prefab = defaultPrefab;
+            }
+
+            Instantiate(prefab);
+        }
+
+        public void SetDefaultPrefab(GameObject prefab)
```
Comment (SimonTheSourcerer) [inline Runtime/CourseController/CourseMenuSpawner.cs:48]: xml doc

File: Runtime/CourseController/Controllers/UIBaseCourseController.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+using UnityEngine;
+
+namespace Innoactive.Creator.UX
+{
+    /// <summary>
+    /// Base course controller which also takes care that a course menu is spawned.
+    /// </summary>
+    public abstract class UIBaseCourseController : BaseCourseController
+    {
+        public abstract string CourseMenuPrefabName { get; }
```
Comment (SimonTheSourcerer) [inline Runtime/CourseController/Controllers/UIBaseCourseController.cs:12]: xml doc

**Review by SimonTheSourcerer — CHANGES_REQUESTED** [review]

**Review by Gusinuhe — COMMENTED** [review]
It feels very weird that `CourseControllerMenu` has way more modifications than `StandaloneCourseMenu`, both scripts are supposed to be identical with the only difference that one uses Unity.UI stuff and the other TextMeshPro, we could even merge them if we rework the `AdvanceCourseController` UI to use TextMeshPro. ⚠️

File: Editor/CourseController/UI/CourseMenuSpawnerEditor.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            defaultPrefabProperty = serializedObject.FindProperty("defaultPrefab");
+            useCustomPrefabProperty = serializedObject.FindProperty("useCustomPrefab");
+            customPrefabProperty = serializedObject.FindProperty("customPrefab");
+        }
+
+        public override void OnInspectorGUI()
+        {
+            GUI.enabled = false;
+            EditorGUILayout.ObjectField("Script", MonoScript.FromMonoBehaviour((CourseMenuSpawner)target), typeof(CourseMenuSpawner), false);
+            EditorGUILayout.ObjectField("Default prefab", defaultPrefabProperty.objectReferenceValue, typeof(GameObject), false);
```
[the remark this answers is not in the document]
Comment (tomwim) [reply Editor/CourseController/UI/CourseMenuSpawnerEditor.cs:34]: I honestly would prefer it for the CourseController, too. It just gives you quicker access to see which prefab is gonna be spawned without looking for it somewhere. But I am fine either way

**Comment by tomwim** [comment]
Unsure what to do with the CourseControllerMenu for now. 

Otherwise ready for re-review @Gusinuhe @SimonTheSourcerer

**Review by SimonTheSourcerer — APPROVED** [review]

**Review by Gusinuhe — APPROVED** [review]

**Pull request merged**